### PR TITLE
[PowerPC] Tune AIX shared library TLS model at function level

### PR DESCRIPTION
**Pull request by orcguru**
Under some circumstance (library loaded with the main program), TLS initial-exec and local-dynamic access model can be interchanged. We could use some simple heuristic to decide use which TLS model when the variable is accessed in which function:
* If there is single TLS variable access in the function, use TLS initial-exec model.
* If there are multiple TLS variable accesses in the function, use TLS local-dynamic model.

### CODE REVIEW

**Comment by orcguru** [comment]
I will push a change to limit the rename to 64bit only.

**Comment by bzEq** [comment]
> so that in following patches we can switch between different TLS models (local-dynamic or initial-exec) for the same TLS GV in different functions within the same module.

Could you please outline what following patches will do?

**Comment by orcguru** [comment]
> Could you please outline what following patches will do?

Under some circumstance (library not loaded by `dlopen`), TLS initial-exec and local-dynamic access model can be interchanged. Some particular TLS variable could be accessed by different functions in the same object file. We could use some simple heuristic to decide use which TLS model when the variable is accessed in which function.

However, to make it able to mix different TLS models towards the same TLS variable in the same object file, we have to create renames. Since it seems to me renaming is fine when there is no mix, this patch is doing the rename for all TLS local-dynamic model accesses.

File: llvm/lib/Target/PowerPC/PPC.td
Comment (bzEq) [inline llvm/lib/Target/PowerPC/PPC.td:333]: This should be an optimization, maybe rename to `aix-shared-library-tls-model-opt`, we perform optimization based on some heuristics.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp
Comment (bzEq) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:3379]:
```suggestion
    SmallPtrSet<const GlobalValue *, 8> TLSGV;
```

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp
Comment (bzEq) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:3380]: One DAG is mapped to a single basic block, I notice your description is about whole function. So the size of `TLSGV` is smaller than what you really want to count.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp
Comment (amy-kwan) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:3410]: Minor nit on comments.
```suggestion
               << " function is using the TLS-LD model for TLS IE/LD accesses.\n");
    Model = TLSModel::LocalDynamic;
  } else if (FuncInfo->isAIXFuncUseTLSIE()) {
    LLVM_DEBUG(
        dbgs() << DAG.getMachineFunction().getName()
               << " function is using the TLS-IE model for TLS IE/LD accesses.\n");
```

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp
Comment (amy-kwan) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:3393]: `TLSGV.size()` is used more than once and can be pulled out.

File: clang/lib/Basic/Targets/PPC.h
Comment (orcguru) [reply clang/lib/Basic/Targets/PPC.h:83]: Hi Kai, Thank you for pointing out this.

I checked the code and it seems this flag is being used by backend in various places (for example: `PPCTargetLowering::LowerGlobalTLSAddressAIX`). You referred to another example that this flag is redundant, may I know the details and I will cross check with that scenario?

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp
Comment (orcguru) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:3380]: Emm, I will update test case to expose this issue. I will propose scan over all instructions in my next update. Thank you!

File: llvm/lib/Target/PowerPC/PPC.td
Comment (orcguru) [reply llvm/lib/Target/PowerPC/PPC.td]: I changed the name to `aix-shared-lib-tls-model-opt`. Hope that is fine.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp
Comment (amy-kwan) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:3372]: This primarily sets the TLS model per function. Not sure if it is possible, but can it/does it make sense for this part to be separated into its own function?

Comment (orcguru) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:3370]: Sure. Created static function guarded by `Subtarget.hasAIXShLibTLSModelOpt()`. Thank you for the suggestion!

Comment (orcguru) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:3370]: Sure. Created static function guarded by `Subtarget.hasAIXShLibTLSModelOpt()`. Thank you for the suggestion!

**Review by amy-kwan — APPROVED** [review]
Thank you for addressing my comment!
I think I do not have any further comments, so unless if anyone has any other comments LGTM.

**Comment by orcguru** [comment]
Since https://github.com/llvm/llvm-project/pull/86641 has been merged, I will continue work on this patch to handle the interactions of these two flags.

Mean while I'm waiting for more review comments from experts. Thank you!

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp
Comment (hubert-reinterpretcast) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:3394]: We should only count the number of `local-dynamic` accesses.

Rationale:
The option can be taken as providing the information that the shared object is loaded with the executable (not `dlopen`ed later). In that case, all variables accessed using the `local-dynamic` model can be accessed using `initial-exec`; however, there would still be variables accessed using `initial-exec` that cannot be accessed using `local-dynamic`.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp
Comment (hubert-reinterpretcast) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:3404]: We should need to set only one thing (to represent "use IE instead of LD"). The threshold number should not be repeated in more than one place.

Additionally, the threshold number may need to be increased (tests via experiments).

**Review by hubert-reinterpretcast — COMMENTED** [review]
See inline comments for rationale re: scoping to LD -> IE only (and not performing IE -> LD).

File: clang/include/clang/Driver/Options.td
Comment (orcguru) [reply clang/include/clang/Driver/Options.td]: I did update the description. Please let me know if another statement is better. Thank you!

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp
Comment (orcguru) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp]: I added target hidden flag `PPCAIXTLSModelOptUseIEForLDLimit` in PPCISelLowering.cpp, and its initial value will be 1. Also updated test case to show the effect of change this parameter.

File: clang/lib/Basic/Targets/PPC.h
Comment (bzEq) [reply clang/lib/Basic/Targets/PPC.h]: The variable defined here is only visible to clang and clang doesn't use this flag further. What backend sees is the feature string and set corresponding variables in backend accordingly.

Comment (bzEq) [reply clang/lib/Basic/Targets/PPC.h]: Code in `PPCTargetLowering::LowerGlobalTLSAddressAIX` like
```
Subtarget.hasAIXSmallLocalExecTLS()
```
which is generated in `PPCGenSubtargetInfo.inc`, whose value is initialized by feature string `aix-small-local-exec-tls` in LLVM IR rather than the variable defined in clang.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp
Comment (amy-kwan) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:3383]: nit on comment:
```suggestion
    // Iterate over all instructions within the current function and collect all TLS
```

**Review by amy-kwan — APPROVED** [review]
LGTM after all the updates. Thanks!

**Pull request merged**
